File: Teledyne DALSA/Sapera/Examples/NET/CameraEvents/VB/CameraEvents.vb
Module CameraEvents

    Sub AcqDeviceCallback(ByVal sender As Object, ByVal args As SapAcqDeviceNotifyEventArgs)
        Dim acqDevice As SapAcqDevice = TryCast(sender, SapAcqDevice)
        Console.WriteLine("AcqDeviceNotify event ""{0}"", Feature = ""{1}""", acqDevice.EventNames(args.EventIndex), acqDevice.FeatureNames(args.FeatureIndex))
    End Sub

    Sub Main(ByVal args As String())
        Dim feature As SapFeature = Nothing
        Dim device As SapAcqDevice = Nothing
        Dim acqParams As New MyAcquisitionParams()
        Dim currentGainDouble As Double = 0
        Dim currentGainInt As Integer = 0
        Dim isGenie As Boolean
        Dim isAvailable As Boolean = False
        Dim isSFNCDeprecated As Boolean = False
        Dim status As Boolean = False

        Console.WriteLine("Sapera Console Cameras Events Example (VB.NET version)" & Chr(10) & "")

        If Not GetOptions(args, acqParams) Then
            Console.WriteLine("" & Chr(10) & "Press any key to terminate" & Chr(10) & "")
            Console.ReadKey()
            Return
        End If

        Dim location As New SapLocation(acqParams.ServerName, acqParams.ResourceIndex)
        ' Create a device object and assign delegate
        device = New SapAcqDevice(location, acqParams.ConfigFileName)
        AddHandler device.AcqDeviceNotify, AddressOf AcqDeviceCallback
        If Not device.Create() Then
            DestroysObjects(device, feature)
            Return
        End If
        ' Create an empty feature object (to receive information)
        feature = New SapFeature(location)
        If Not feature.Create() Then
            DestroysObjects(device, feature)
            Return
        End If
        ' Browse event list
        Dim numEvents As Integer = device.EventCount
        Console.WriteLine("Available events are:" & Chr(10) & "")
        For eventIndex As Integer = 0 To numEvents - 1
            Console.WriteLine("  " + device.EventNames(eventIndex))
        Next
        Console.WriteLine()

        ' Register event by name
        If Not device.IsEventEnabled("Feature Value Changed") Then
            device.EnableEvent("Feature Value Changed")
        End If

        Dim modelName As String = ""
        status = device.IsFeatureAvailable("DeviceModelName")
        If (status) Then
            status = device.IsFeatureAvailable("FrameRate")
            device.GetFeatureValue("DeviceModelName", modelName)
            If (status And modelName.Contains("Genie")) Then
                isGenie = True
            End If
        End If

        isAvailable = device.IsFeatureAvailable("Gain")
        If (isAvailable) Then
            device.GetFeatureInfo("Gain", feature)
            GoTo Gain_available
        End If
        isAvailable = device.IsFeatureAvailable("GainRaw")
        If (isAvailable) Then
            device.GetFeatureInfo("GainRaw", feature)
            isSFNCDeprecated = True
        Else
            GoTo Feature_Not_Available
        End If

Gain_Available:
        If (isGenie Or isSFNCDeprecated) Then
            Dim gainMin As Integer = 0, gainExponent As Integer = 0, gainMax As Integer = 0, gainValue As Integer = 0, powValue As Double = 0
            feature.GetValueMax(gainMax)
            feature.GetValueMin(gainMin)
            gainExponent = feature.SiToNativeExp10
            powValue = Convert.ToDouble(Math.Pow(CSng(10), -gainExponent))

            ' Get current Gain value in camera
            If (isSFNCDeprecated) Then
                device.GetFeatureValue("GainRaw", currentGainInt)
            Else
                device.GetFeatureValue("Gain", currentGainInt)
            End If

            Console.WriteLine("" & Chr(10) & "Current gain value is " + Convert.ToString(currentGainInt) + "" & Chr(10) & "")

            gainValue = Convert.ToInt32(gainMax * powValue)
            If (gainValue = currentGainInt) Then
                gainValue = Convert.ToInt32(gainMin * powValue)
            End If

            If (isSFNCDeprecated) Then
                device.SetFeatureValue("GainRaw", gainMax)
            Else
                device.SetFeatureValue("Gain", gainMax)
            End If
            Console.WriteLine("" & Chr(10) & "Set Gain value to " + Convert.ToString(gainValue) + "" & Chr(10) & "")


        Else
            Dim gainMin As Double = 0, gainExponent As Integer = 0, gainMax As Double = 0, gainValue As Double = 0, powValue As Double = 0
            feature.GetValueMax(gainMax)
            feature.GetValueMin(gainMin)
            gainExponent = feature.SiToNativeExp10
            powValue = Convert.ToDouble(Math.Pow(CSng(10), -gainExponent))

            ' Get current Gain value in camera
            device.GetFeatureValue("Gain", currentGainDouble)
            Console.WriteLine("" & Chr(10) & "Current gain value is " + String.Format("{0:n2}", currentGainDouble) + "" & Chr(10) & "")

            'Set gain to max. if it's already at max, set it to min.
            'The 0.001 adjustment is to avoid rounded numbers larger/lower
            'than max/min which would trigger errors on the camera.
            gainValue = gainMax * powValue - 0.001
            If (gainValue = currentGainDouble) Then
                gainValue = gainMin * powValue + 0.001
            End If

            
            device.SetFeatureValue("Gain", gainValue)
            Console.WriteLine("" & Chr(10) & "Set Gain value to " + String.Format("{0:n2}", gainValue) + "" & Chr(10) & "")
        End If

        ' Unregister event by name
        If device.IsEventEnabled("Feature Value Changed") Then
            device.DisableEvent("Feature Value Changed")
        End If
Feature_Not_Available:

        Console.WriteLine("Press any key to terminate.")
        Console.ReadKey()

        ' Set Gain value to old value
        If (isAvailable) Then
            If (isGenie) Then
                device.SetFeatureValue("Gain", currentGainInt)
            ElseIf (isSFNCDeprecated) Then
                device.SetFeatureValue("GainRaw", currentGainInt)
            Else
                device.SetFeatureValue("Gain", currentGainDouble)
            End If
            Console.WriteLine("" & Chr(10) & "Set Gain to old value.")
        End If

        DestroysObjects(device, feature)
    End Sub

    Function GetOptions(ByVal args As String(), ByVal acparams As MyAcquisitionParams) As Boolean
        ' Check if arguments were passed
        If args.Length > 1 Then
            Return ExampleUtils.GetAcqDeviceOptionsFromCommandLine(args, acparams)
        Else
            Return ExampleUtils.GetCorAcqDeviceOptionsFromQuestions(acparams, False)
        End If
    End Function

    Sub DestroysObjects(ByVal acq As SapAcqDevice, ByVal feat As SapFeature)
        If acq IsNot Nothing Then
            acq.Destroy()
            acq.Dispose()
        End If
        If feat IsNot Nothing Then
            feat.Destroy()
            feat.Dispose()
        End If
        
    End Sub


End Module

File: Teledyne DALSA/Sapera/Examples/NET/CameraEvents/CSharp/CameraEvents.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;
using System.Threading;

using DALSA.SaperaLT.SapClassBasic;
using DALSA.SaperaLT.Examples.NET.Utils;

namespace DALSA.SaperaLT.Examples.NET.CSharp.CameraEvents
{
    class Program
    {
        static void AcqDeviceCallback(Object sender, SapAcqDeviceNotifyEventArgs args)
        {
            SapAcqDevice acqDevice = sender as SapAcqDevice;
            Console.WriteLine("AcqDeviceNotify event \"{0}\", Feature = \"{1}\"",
                acqDevice.EventNames[args.EventIndex], acqDevice.FeatureNames[args.FeatureIndex]);
        }

        static void Main(string[] args)
        {
            SapFeature feature = null;
            SapAcqDevice device = null;
            MyAcquisitionParams acqParams = new MyAcquisitionParams();
            
            Console.WriteLine("Sapera Console Cameras Events Example (C# version)\n");

            if (!GetOptions(args, acqParams))
            {
                Console.WriteLine("\nPress any key to terminate\n");
                Console.ReadKey();
                return;
            }

            SapLocation location = new SapLocation(acqParams.ServerName, acqParams.ResourceIndex);
            // Create a camera object and assign delegate
            device = new SapAcqDevice(location, acqParams.ConfigFileName);
            device.AcqDeviceNotify += new SapAcqDeviceNotifyHandler(AcqDeviceCallback);
            if (!device.Create())
            {
                DestroysObjects(device,feature);
                return;
            }
            // Create an empty feature object (to receive information)
            feature = new SapFeature(location);
            if (!feature.Create())
            {
                DestroysObjects(device,feature);
                return;
            }
            // Get the number of features provided by the camera
            int featureCount = device.FeatureCount;         
            // Browse event list
            int numEvents = device.EventCount;       
            Console.WriteLine("Available events are:\n");
            for (int eventIndex = 0; eventIndex < numEvents; eventIndex++)
            {             
                Console.WriteLine("  " + device.EventNames[eventIndex]);
            }
            Console.WriteLine();

            // Register event by name
            if (!device.IsEventEnabled("Feature Value Changed"))
                 device.EnableEvent("Feature Value Changed");

            // Modified a feature (Will trigger callback function)
            double currentGainDouble =0;
            int currentGainInt=0;
            Boolean isGenie = false, isAvailable = false, isSFNCDeprecated = false, status = false;
            string modelName;

            


            status = device.IsFeatureAvailable("DeviceModelName");
            if (status)
            {
                status = device.IsFeatureAvailable("FrameRate");
                device.GetFeatureValue("DeviceModelName", out modelName);
                if (status && modelName.Contains("Genie"))
                {
                    isGenie = true;
                }
            }

            status = false;
            isAvailable = device.IsFeatureAvailable("Gain");
            if (isAvailable)
                device.GetFeatureInfo("Gain", feature);
            else if (isAvailable = device.IsFeatureAvailable("GainRaw"))
            {
                device.GetFeatureInfo("GainRaw", feature);
                isSFNCDeprecated = true;
            }
            else
            {
                goto Feature_Not_Available;
            }

            if (isGenie || isSFNCDeprecated)
            {
                int gainMin = 0, gainExponent = 0, gainMax = 0, gainValue = 0;
                double powValue = 0;
                feature.GetValueMax(out gainMax);
                feature.GetValueMin(out gainMin);
                gainExponent = feature.SiToNativeExp10;
                powValue = Convert.ToDouble(Math.Pow((float)10, -gainExponent));
                
                // Get current Gain value in camera
                if (isSFNCDeprecated)
                    device.GetFeatureValue("GainRaw", out currentGainInt);
                else
                    device.GetFeatureValue("Gain", out currentGainInt);

                Console.WriteLine("\nCurrent gain value is " + Convert.ToString(currentGainInt) + "\n");

                gainValue = Convert.ToInt32(gainMax * powValue);
                if(currentGainInt == gainValue)
                    gainValue = Convert.ToInt32(gainMin * powValue);

                if(isSFNCDeprecated)
                    device.SetFeatureValue("GainRaw", gainMax);
                else
                    device.SetFeatureValue("Gain", gainMax);
                Console.WriteLine("\nSet Gain value to " + Convert.ToString(gainValue) + "\n");
                
            }
            else 
            {
                double gainMin = 0, gainMax = 0, gainValue = 0, powValue = 0;
                int gainExponent = 0;
                feature.GetValueMax(out gainMax);
                feature.GetValueMin(out gainMin);
                gainExponent = feature.SiToNativeExp10;
                powValue = Convert.ToDouble(Math.Pow((float)10, -gainExponent));
                
                // Get current Gain value in camera
                device.GetFeatureValue("Gain", out currentGainDouble);
                Console.WriteLine("\nCurrent gain value is " + String.Format("{0:0.00}",currentGainDouble) + "\n");

                //Set gain to max. if it's already at max, set it to min.
                //The 0.001 adjustment is to avoid rounded numbers larger/lower
                //than max/min which would trigger errors on the camera.
                gainValue = gainMax * powValue - 0.001;
                if (currentGainDouble == gainValue)
                    gainValue = gainMin * powValue + 0.001;

                device.SetFeatureValue("Gain", gainValue);
                Console.WriteLine("\nSet Gain value to " + String.Format("{0:0.00}", gainValue) + "\n");
            }
            
Feature_Not_Available:
            // Unregister event by name
            if (device.IsEventEnabled("Feature Value Changed"))
               device.DisableEvent("Feature Value Changed");

            Console.WriteLine("Press any key to terminate.");
            Console.ReadKey();

            if (isAvailable)
            {
                // Set Gain value to old value
                if (isGenie)
                    device.SetFeatureValue("Gain", currentGainInt);
                else if (isSFNCDeprecated)
                    device.SetFeatureValue("GainRaw", currentGainInt);
                else
                    device.SetFeatureValue("Gain", currentGainDouble);
                Console.WriteLine("\nSet Gain to old value.");
            }

            DestroysObjects(device, feature);
        }

        static bool GetOptions(string[] args, MyAcquisitionParams acparams)
        {
            // Check if arguments were passed
            if (args.Length > 1)
                return ExampleUtils.GetAcqDeviceOptionsFromCommandLine(args, acparams);
            else
                return ExampleUtils.GetCorAcqDeviceOptionsFromQuestions(acparams, false);
        }

        static void DestroysObjects(SapAcqDevice acq, SapFeature feat)
        {
            if (acq != null)
            {
                acq.Destroy();
                acq.Dispose();
            }
            if (feat != null)
            {
                feat.Destroy();
                feat.Dispose();
            }       
           
        }

    }
}
